### ISSUE TITLE
HiRAG on ArangoDB — PRD (v1.1, consolidated)

### ISSUE THREAD

**Issue by r3d91ll** [r3d91ll/HADES-Lab#19]
# [HiRAG](https://arxiv.org/abs/2503.10150) on ArangoDB — PRD (v1.1, consolidated)

## 0) Executive Summary

**Product**: **HiRAG‑Arango** — Hierarchical Retrieval‑Augmented Generation on ArangoDB.
**Goal**: Implement HiRAG’s three‑level retrieval (Local ↔ Bridge ↔ Global) on ArangoDB to accelerate theory‑→practice discovery over the arXiv corpus, instrumented with the Conveyance Framework.

**Why now**: ArangoDB’s multi‑model (documents + graphs) and AQL make it a strong substrate for hierarchical KGs while keeping retrieval latency tractable at arXiv scale.

---

## 1) Objectives & KPIs (Conveyance‑aligned)

* **Primary Objective**: Deliver three‑level HiRAG retrieval with measurable improvements vs baseline RAG.
* **Key KPIs** (target launch SLOs):

  * **W (Semantic Quality)**: Top‑k semantic relevance > **0.85**.
  * **R (Graph Connectivity)**: **Bridge discovery rate ≥ 0.80** with avg path length < **4 hops**.
  * **H/T (Efficiency)**: P95 query latency < **2.0 s**; throughput ≥ **100 QPM**.
  * **Ctx (L/I/A/G)**: Context completeness ≥ **0.80**; local↔global coherence ≥ **0.75**.
  * **IR Validation**: Observe **α ≈ 1.6** amplification in retrieval quality; verify zero‑propagation behavior when any hierarchy level is missing.

---

## 2) Scope & Non‑Goals

* **In**: ArangoDB schema, hierarchical index (HiIndex), weighted path retrieval (HiRetrieval), clustering + weighting strategies, conveyance scoring, validation harness, perf strategies.
* **Out** (v1): RL/agentic traversal learning; cross‑DB federation; UI dashboards beyond basic metrics.

---

### 3.1 Collections & Indices

```aql
// Document collections
papers          // raw arXiv papers & metadata
entities        // extracted entities (paper|concept|method|person)
clusters        // hierarchy summary nodes (level 1 & 2)

// Edge collections
relations       // entity↔entity (cites|implements|extends|summarizes)
cluster_edges   // member↔cluster and cluster↔cluster links

// Optional ops/analytics
query_logs, bridge_cache, weight_config
```

**Representative docs**

```jsonc
// papers
{
  "_key": "arxiv_id",
  "title": "...",
  "abstract": "...",
  "content": "...",
  "categories": ["cs.AI","cs.LG"],
  "date": "ISO_date",
  "embeddings": [ ... ],
  "conveyance_score": 0.0
}

// entities
{
  "_key": "entity_id",
  "name": "multi-head attention",
  "type": "concept",
  "layer": 0,              // 0=base, 1=summary, 2=super-summary
  "cluster_id": "...",
  "description": "...",
  "semantic_embedding": [ ... ]
}

// relations (edges)
{
  "_from": "entities/e1",
  "_to":   "entities/e2",
  "type":  "cites",        // cites|implements|extends|summarizes
  "weight": 0.0,
  "layer_bridge": true,     // crosses hierarchy levels?
  "created": "ISO_date",
  "embedding_similarity": 0.0,
  "conveyance_weight": 0.0  // materialized edge weight
}
```

### 3.2 Traversal Patterns

```aql
// Find bridge paths between local and global knowledge
FOR start IN local_entities
  FOR end IN global_entities
    FOR path IN OUTBOUND SHORTEST_PATH start TO end relations
      FILTER path.edges[*].layer_bridge == true
      RETURN path
```

---

### 4.1 Clustering Strategy (Level 0 → Level 1 → Level 2)

**Hybrid clustering** combining **semantic** and **structural** signals.

1. **Embedding‑based** topic clusters (Level 0 → Level 1)

* Source: `entities.layer=0.semantic_embedding` (e.g., Jina v4 / ModernBERT).
* Method: **HDBSCAN** (cosine metric), `min_cluster_size≈5`, `min_samples≈3`.

2. **Citation community detection** (graph‑structural)

* Build 1–2 hop citation neighborhoods; apply **Louvain** (or Leiden) for communities.

3. **Fusion**

* Merge with weights: `embedding: 0.6`, `citation: 0.4`.
* Break ties by maximizing intra‑cluster cosine similarity and minimizing inter‑cluster conductance.

4. **Level 1 → Level 2** (super‑clusters)

* Create **meta‑embeddings** from Level 1 summaries (mean pooling + keyphrase TF‑IDF).
* Apply **agglomerative clustering** (Ward); target `n≈√(level1_count)`.

### 4.2 Cluster Summaries

* Extract **key concepts** (TF‑IDF / YAKE) from member descriptions.
* Generate \~150–200 token **LLM summaries**.
* Store as `clusters` with `layer∈{1,2}`, `member_count`, and `semantic_embedding` of the summary.

---

### 5.1 Three‑Level Flow

1. **Local** (entity‑level match)

```aql
FOR e IN entities
  FILTER e.layer == 0
  FILTER ANALYZER(e.description, "text_en") LIKE @query
  RETURN e
```

2. **Global** (cluster expansion)

```aql
FOR e IN @local
  FOR c IN OUTBOUND e cluster_edges
    FILTER c.layer >= 1
    RETURN c
```

3. **Bridge** (weighted shortest paths)

* Multi‑factor **edge weights** (see §6) and `SHORTEST_PATH ... OPTIONS { weightAttribute: 'conveyance_weight' }`.

### 5.2 Engine Skeleton

```js
class HiRetrievalEngine {
  async retrieveLocal(query) { /* AQL above */ }
  async retrieveGlobal(local) { /* AQL above */ }
  async retrieveBridge(local, global) { /* weighted shortest paths */ }
}
```

---

### 6.1 Edge Weight Schema

* **Base by type**: `cites:0.8`, `implements:0.9`, `extends:0.7`, `summarizes:0.6`
* **Modifiers**:

  * **Temporal decay**: `exp(-age_days/365)`
  * **Embedding similarity**: `sim²` (emphasize higher similarity)
  * **Cross‑layer bonus**: `×1.5` if `layer_bridge=true`

**Materialization**: Precompute `conveyance_weight` on `relations` edges; periodically refresh.

### 6.2 Weighted Shortest Path (AQL sketch)

```aql
FOR s IN @locals
  FOR t IN @globals
    FOR p IN OUTBOUND s relations
      OPTIONS { weightAttribute: 'conveyance_weight', defaultWeight: 0.1 }
      SHORTEST_PATH TO t
      LIMIT 5
      RETURN {
        path: p,
        totalWeight: SUM(FOR e IN p.edges RETURN e.conveyance_weight)
      }
```

### 6.3 Online Weight Adaptation

* Collect user feedback (clicks, success flags) → **gradient update** for `weight_config`.
* Low‑risk learning rate (e.g., 0.01) with guardrails.

---

### 7.1 Concrete Metrics

* **W — Semantic Quality (0–1)**: weighted cosine similarity of retrieved entities vs query embedding; weights via PageRank‑style importance.
* **R — Graph Connectivity (0–1)**: combine

  * normalized inverse path length,
  * avg edge weight over path,
  * cross‑layer bridge ratio.
* **H/T — Efficiency**: fix **H** by model class (baseline 0.8); capture **T** = observed retrieval latency.
* **Ctx = ¼(L + I + A + G)**:

  * **L (Local coherence)**: intra‑cluster similarity of local results.
  * **I (Instruction fit)**: alignment of bridge evidence with the query intent (keyword overlap / classifier score).
  * **A (Actionability)**: presence of implementations / methods / examples.
  * **G (Grounding)**: citations, datasets, code links present.
* **α**: start **1.6**; track empirically.

---

### 8.1 Pre‑computation & Caching

* **bridge\_cache**: precompute hot pair bridges from `query_logs` (freq>10, TTL=7d).
* Materialize `conveyance_weight` and per‑cluster **centroids**.

### 8.2 Query‑time Guards

* BFS with **limits**; cap candidate locals/globals; prune by similarity/topology.
* Parallelize local/global; short‑circuit on high‑confidence paths.

### 8.3 Sharding & Indexing

* Shard `entities`, `relations` by hash of `_key` or community id; persistent indexes on `type`, `layer`, `cluster_id`, and `relations.type`.

---

### 9.1 Ground Truth (Theory→Practice pairs)

* Curate ≥ **500** arXiv→GitHub pairs (e.g., 1706.03762 → HF/Transformers, fairseq, t2t).
* For each: expected **bridge entities** (e.g., transformer, multi‑head attention, positional encoding).

### 9.2 Metrics

* **Bridge Discovery Rate** (hit if any retrieved path contains expected bridge entities).
* **Context Coherence**; **Latency**; **Precision\@k** vs Baseline RAG (no hierarchy).

### 9.3 Targets

* Baseline: `bridgeDiscovery≈0.45`, `context≈0.65`, `latency≈1.2s`.
* HiRAG Goal: `bridgeDiscovery≥0.80`, `context≥0.85`, `latency≤1.8s`.

---

### Phase 1 — Foundation (Weeks 1–2)

* Provision ArangoDB (cluster or single‑node), create collections/indexes.
* Bulk import **papers**; bootstrap **entities** (DocProc + NER); initial **relations**.

### Phase 2 — HiIndex (Weeks 3–4)

* Run hybrid clustering; create **clusters** (L1/L2) + **cluster\_edges**.
* Generate cluster summaries + embeddings; mark `layer_bridge` edges.

### Phase 3 — HiRetrieval (Weeks 5–6)

* Implement local/global/bridge queries; materialize `conveyance_weight`.
* Integrate conveyance scoring; add small feedback loop for weights.

### Phase 4 — Integration & Testing (Weeks 7–8)

* LLM tie‑in for answer synthesis; perf benchmarking; validation harness over GT set.

**MVP Exit Criteria**

* Three‑level retrieval on **10k** paper subset; bridge discovery working; P95 latency < **2 s**.

**Full Release** (Week 12)

* Full corpus indexing; ≥ **25%** uplift vs baseline on bridge discovery; α observed ≈ **1.6**.

---

## 11) Resource Plan

```yaml
ArangoDB Cluster:
  nodes: 3
  memory: 32GB per node
  storage: 1TB SSD per node
  cpu: 16 cores per node

Application Layer:
  servers: 2
  memory: 16GB per server
  cpu: 8 cores per server
```

**Storage estimate**: Papers \~100GB; Entities \~50GB; Relations \~50GB; Indexes \~50GB ⇒ **\~250GB** (pre‑replication).

---

## 12) Risks & Mitigations

* **Traversal latency @50M+ edges** → Precompute hot bridges; prune; materialize weights; shard wisely.
* **Hierarchy quality drift** → Scheduled reclustering; unit tests on cluster cohesion; human spot‑checks.
* **Weight overfitting** → Bound learning rate; A/B on held‑out pairs.

---

### A) AQL/JS Snippets

```js
// Conveyance calculator sketch
function calculateConveyance(ctx){
  const {entities, paths, retrievalTime} = ctx;
  const W = calcSemanticQuality(entities);
  const R = calcGraphConnectivity(paths);
  const H = 0.8; const T = retrievalTime;
  const L = calcLocalCoherence(entities);
  const I = calcInstructionFit(paths);
  const A = calcActionability(ctx);
  const G = calcGrounding(entities);
  const Ctx = 0.25*(L+I+A+G);
  const alpha = 1.6;
  return (W*R*H/T) * Math.pow(Ctx, alpha);
}
```

```aql
// Weighted shortest paths (materialized weights)
FOR s IN @locals
  FOR t IN @globals
    FOR p IN OUTBOUND s relations
      OPTIONS { weightAttribute: 'conveyance_weight', defaultWeight: 0.1 }
      SHORTEST_PATH TO t
      LIMIT 5
      RETURN p
```

---

## 14) Change Log

* **v1.1**: Consolidated PRD + Addendum; added hybrid clustering, edge weighting, concrete conveyance metrics, perf/cache strategy, validation dataset.
* **v1.0**: Initial PRD draft.
